openff/toolkit/topology/topology.py

```diff
@@ -441,6 +444,37 @@ def _initialize(self):
         self._molecules = list()
         self._cached_chemically_identical_molecules = None
 
+    def __iadd__(self, other):
+        """Add two Topology objects in-place.
+
+        This method has the following effects:
+        * The cache (_cached_chemically_identical_molecules) is reset
+        * The constrained atom pairs (Topology.constrained_atom_pairs) is reset
+        * Box vectors are **not** updated.
+
+        """
+        if self.aromaticity_model != other.aromaticity_model:
+            raise InvalidAromaticityModelError(
+                "Mismatch in aromaticity models. Trying to add a Topology with aromaticity model "
+                f"{other.aromaticity_model} to a Topology with aromaticity model "
+                f"{self.aromaticity_model}"
+            )
+
+        self._constrained_atom_pairs = dict()
```

Comment:
(blocking) Hmmm, after some thought I think that resetting constraints is a big footgun waiting to go off. While I'd LIKE to think that constraints live entirely in post-parameterization-land (and therefore wouldn't be important to keep track of) we DO currently offer some guarantee that a constraint defined directly on the Topology before parameterization will [be put into the resulting system](https://github.com/openforcefield/openff-toolkit/blob/921fbe2d48f3069adc3af6e919349dee6deefa26/openff/toolkit/typing/engines/smirnoff/parameters.py#L2841-L2856). So there could be workflows out there that define constraints ahead of time, and this would be a silent total break for them.

So, for this PR to be accepted, it would need to keep track of the constraints during topology addition. While the `+` operator should be magic, it shouldn't be quite _this_ magic.

Reply:
Added in https://github.com/openforcefield/openff-toolkit/pull/1194/commits/3452b2d137f065521b4d5ca228bc5be3936944cd